Failed tests (10813): TestGetZoneConfig

The following test appears to have failed:

[#10813](https://circleci.com/gh/cockroachdb/cockroach/10813):

```
Wr@0.000000000,0  127.0.0.1:0  10:07:47.004077  10.41µs     ·meta descriptor lookup       kv/dist_sender.go:562
Wr@0.000000000,0  127.0.0.1:0  10:07:47.004098  1.87545ms   ·sending RPC                  kv/dist_sender.go:422
Wr@0.000000000,0  127.0.0.1:0  10:07:47.004143  0           ··sending to 127.0.0.1:45380  rpc/send.go:154

I0108 10:07:49.317899 860 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- FAIL: TestGetZoneConfig (17.33s)
	config_test.go:192: failed to get latest system config: sql/config_test.go:77: condition failed to evaluate true within 10s
=== RUN   TestDatabaseDescriptor
I0108 10:07:49.321931 860 rpc/clock_offset.go:153  monitoring cluster offset
I0108 10:07:49.327860 860 raft/raft.go:446  [group 1] 1 became follower at term 5
I0108 10:07:49.327920 860 raft/raft.go:234  [group 1] newRaft 1 [peers: [1], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0108 10:07:49.327938 860 raft/raft.go:526  [group 1] 1 is starting a new election at term 5
I0108 10:07:49.327954 860 raft/raft.go:459  [group 1] 1 became candidate at term 6
I0108 10:07:49.327968 860 raft/raft.go:508  [group 1] 1 received vote from 1 at term 6
I0108 10:07:49.327993 860 raft/raft.go:487  [group 1] 1 became leader at term 6
I0108 10:07:49.328739 860 storage/replica_command.go:1198  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC +1452247670.328s
--
--- PASS: TestValidateTableDesc (0.00s)
=== RUN   TestColumnTypeSQLString
--- PASS: TestColumnTypeSQLString (0.01s)
=== RUN   TestInitialKeys
--- PASS: TestInitialKeys (0.00s)
FAIL
FAIL	github.com/cockroachdb/cockroach/sql	64.467s
=== RUN   TestNoLinkForbidden
--- PASS: TestNoLinkForbidden (0.12s)
=== RUN   TestDates
I0108 10:07:31.869088 854 rpc/clock_offset.go:153  monitoring cluster offset
I0108 10:07:31.873059 854 raft/raft.go:446  [group 1] 1 became follower at term 5
I0108 10:07:31.873151 854 raft/raft.go:234  [group 1] newRaft 1 [peers: [1], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0108 10:07:31.873227 854 raft/raft.go:526  [group 1] 1 is starting a new election at term 5
I0108 10:07:31.873261 854 raft/raft.go:459  [group 1] 1 became candidate at term 6
I0108 10:07:31.873296 854 raft/raft.go:508  [group 1] 1 received vote from 1 at term 6
I0108 10:07:31.873456 854 raft/raft.go:487  [group 1] 1 became leader at term 6

```
Please assign, take a look and update the issue accordingly.

--- a/config/config.go
+++ b/config/config.go
@@ -18,6 +18,7 @@ package config
 
 import (
 	"bytes"
+	"crypto/sha1"
 	"fmt"
 	"sort"
 
@@ -115,6 +116,18 @@ func ObjectIDForKey(key roachpb.RKey) (uint32, bool) {
 	// Consume first encoded int.
 	_, id64, err := encoding.DecodeUvarint(key)
 	return uint32(id64), err == nil
+}
+
+// Hash returns a SHA1 hash of the SystemConfig contents.
+func (s SystemConfig) Hash() []byte {
+	sha := sha1.New()
+	for _, kv := range s.Values {
+		_, _ = sha.Write(kv.Key)
+		// There are all kinds of different types here, so we can't use the typed
+		// getters.
+		_, _ = sha.Write(kv.Value.RawBytes)
+	}
+	return sha.Sum(nil)
 }
 
 // GetValue searches the kv list for 'key' and returns its

--- a/gossip/infostore.go
+++ b/gossip/infostore.go
@@ -52,6 +52,10 @@ type infoStore struct {
 	NodeAddr  util.UnresolvedAddr `json:"-"`               // Address of node owning this info store: "host:port"
 	nodes     map[int32]*Node     // Per-node information for gossip peers
 	callbacks []*callback
+
+	callbackMu     sync.Mutex // Serializes callbacks
+	callbackWorkMu sync.Mutex // Protects callbackWork
+	callbackWork   []func()
 }
 
 var monoTime struct {
@@ -101,8 +105,8 @@ func (is *infoStore) String() string {
 }
 
 // newInfoStore allocates and returns a new infoStore.
-func newInfoStore(nodeID roachpb.NodeID, nodeAddr util.UnresolvedAddr) infoStore {
-	return infoStore{
+func newInfoStore(nodeID roachpb.NodeID, nodeAddr util.UnresolvedAddr) *infoStore {
+	return &infoStore{
 		Infos:    make(infoMap),
 		NodeID:   nodeID,
 		NodeAddr: nodeAddr,
@@ -237,10 +241,33 @@ func (is *infoStore) processCallbacks(key string, content roachpb.Value) {
 		}
 	}
 
-	// Run callbacks in a goroutine to avoid mutex reentry.
-	go func() {
+	// Add the matching callbacks to the callback work list.
+	f := func() {
 		for _, method := range matches {
 			method(key, content)
+		}
+	}
+	is.callbackWorkMu.Lock()
+	is.callbackWork = append(is.callbackWork, f)
+	is.callbackWorkMu.Unlock()
+
+	// Run callbacks in a goroutine to avoid mutex reentry. We also guarantee
+	// callbacks are run in order such that if a key is updated twice in
+	// succession, the second callback will never be run before the first.
+	go func() {
+		// Grab the callback mutex to prevent another goroutine from invoking
+		// callbacks.
+		is.callbackMu.Lock()
+		defer is.callbackMu.Unlock()
+
+		// Grab and execute the list of work.
+		is.callbackWorkMu.Lock()
+		work := is.callbackWork
+		is.callbackWork = nil
+		is.callbackWorkMu.Unlock()
+
+		for _, w := range work {
+			w()
 		}
 	}()
 }

--- a/gossip/infostore_test.go
+++ b/gossip/infostore_test.go
@@ -168,7 +168,7 @@ func TestAddInfoSameKeyDifferentHops(t *testing.T) {
 }
 
 // Helper method creates an infostore with 10 infos.
-func createTestInfoStore(t *testing.T) infoStore {
+func createTestInfoStore(t *testing.T) *infoStore {
 	is := newInfoStore(1, emptyAddr)
 
 	for i := 0; i < 10; i++ {

--- a/gossip/server.go
+++ b/gossip/server.go
@@ -44,7 +44,7 @@ type clientInfo struct {
 // newly arrived information on a periodic basis.
 type server struct {
 	mu       sync.Mutex                // Protects the fields below
-	is       infoStore                 // The backing infostore
+	is       *infoStore                // The backing infostore
 	closed   bool                      // True if server was closed
 	incoming nodeSet                   // Incoming client node IDs
 	lAddrMap map[string]clientInfo     // Incoming client's local address -> client's node info

--- a/storage/replica.go
+++ b/storage/replica.go
@@ -21,7 +21,6 @@ package storage
 
 import (
 	"bytes"
-	"crypto/sha1"
 	"errors"
 	"fmt"
 	"math/rand"
@@ -1651,7 +1650,8 @@ func (r *Replica) maybeGossipSystemConfig() {
 	}
 
 	if log.V(1) {
-		log.Infoc(ctx, "gossiping system config from store %d, range %d", r.store.StoreID(), r.RangeID)
+		log.Infoc(ctx, "gossiping system config from store %d, range %d, hash %x",
+			r.store.StoreID(), r.RangeID, hash)
 	}
 
 	cfg := &config.SystemConfig{Values: kvs}
@@ -1920,19 +1920,8 @@ func (r *Replica) loadSystemDBSpan() ([]roachpb.KeyValue, []byte, error) {
 		r.handleSkippedIntents(intents)
 		return nil, nil, errSystemDBIntent
 	}
-	sha := sha1.New()
 	kvs := br.Responses[0].GetInner().(*roachpb.ScanResponse).Rows
-	for _, kv := range kvs {
-		if _, err := sha.Write(kv.Key); err != nil {
-			return nil, nil, err
-		}
-		// There are all kinds of different types here, so we can't use the
-		// typed getters.
-		if _, err := sha.Write(kv.Value.RawBytes); err != nil {
-			return nil, nil, err
-		}
-	}
-	return kvs, sha.Sum(nil), err
+	return kvs, config.SystemConfig{Values: kvs}.Hash(), err
 }
 
 // maybeAddToSplitQueue checks whether the current size of the range

